flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStoreTest.java

```diff
@@ -0,0 +1,306 @@
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@link DefaultJobGraphStore} with {@link TestingJobGraphStoreWatcher}, {@link TestingStateHandleStore}
+ * and {@link TestingJobGraphListener}.
+ */
+public class DefaultJobGraphStoreTest extends TestLogger {
+
+	private final JobGraph testingJobGraph = new JobGraph();
+	private final long timeout = 3 * 1000;
+
+	private TestingStateHandleStore.Builder<JobGraph> builder;
+	private TestingRetrievableStateStorageHelper<JobGraph> jobGraphStorageHelper;
+	private TestingJobGraphStoreWatcher testingJobGraphStoreWatcher = new TestingJobGraphStoreWatcher();
+	private TestingJobGraphListener testingJobGraphListener = new TestingJobGraphListener();
+
+	@Before
+	public void setup() {
+		builder = TestingStateHandleStore.builder();
+		jobGraphStorageHelper = new TestingRetrievableStateStorageHelper<>();
+	}
+
+	@After
+	public void teardown() {
+		if (testingJobGraphStoreWatcher != null) {
+			testingJobGraphStoreWatcher.stop();
+		}
+	}
+
+	@Test
+	public void testRecoverJobGraph() throws Exception {
+		final RetrievableStateHandle<JobGraph> stateHandle = jobGraphStorageHelper.store(testingJobGraph);
+		final TestingStateHandleStore<JobGraph> stateHandleStore = builder
+			.setGetFunction(ignore -> stateHandle)
+			.build();
+
+		final JobGraphStore jobGraphStore = createAndStartJobGraphStore(stateHandleStore);
+
+		final JobGraph recoveredJobGraph = jobGraphStore.recoverJobGraph(testingJobGraph.getJobID());
+		assertThat(recoveredJobGraph, is(notNullValue()));
+		assertThat(recoveredJobGraph.getJobID(), is(testingJobGraph.getJobID()));
+
+		jobGraphStore.stop();
+	}
+
+	@Test
+	public void testRecoverJobGraphWhenNotExist() throws Exception {
+		final TestingStateHandleStore<JobGraph> stateHandleStore = builder
+			.setGetFunction(ignore -> {
+				throw new StateHandleStore.NotExistException("Not exist exception.");
+			})
+			.build();
+
+		final JobGraphStore jobGraphStore = createAndStartJobGraphStore(stateHandleStore);
+
+		final JobGraph recoveredJobGraph = jobGraphStore.recoverJobGraph(testingJobGraph.getJobID());
+		assertThat(recoveredJobGraph, is(nullValue()));
+	}
+
+	@Test
+	public void testRecoverJobGraphExceptionForwarding() throws Exception {
+		final TestingStateHandleStore<JobGraph> stateHandleStore = builder
+			.setGetFunction(ignore -> {
+				throw new FlinkException("Other exception.");
+			})
+			.build();
+
+		final JobGraphStore jobGraphStore = createAndStartJobGraphStore(stateHandleStore);
+
+		final String expectedErrorMsg = "Could not retrieve the submitted job graph state handle " +
+			"for " + testingJobGraph.getJobID() + " from the submitted job graph store";
+		try {
+			jobGraphStore.recoverJobGraph(testingJobGraph.getJobID());
+			fail("Exception should be thrown");
+		} catch (Exception ex) {
+			assertThat(ex.getMessage(), containsString(expectedErrorMsg));
+		}
```

Comment:
I think we can define a custom exception type, and use `@Test(expect = xxx)` for the validation

Reply:
I think we do not need to check the `Exception` type here. So I will use the new introduced `FlinkMatcher#containsMessage` to check the exception.